Move to nose2 only

klecki opened the pull request:
Use nose2 as the only testing framework. Drop nose.

<!---
Thank you for contributing to NVIDIA DALI! If you haven't yet,
please read the contributing guidelines in the CONTRIBUTING.md file.

We need a few more information from you to proceed.
Please fill the relevant sections in this PR template.

Fields in the Checklist section can be marked after you create and save the Pull Request.
--->


## Category: **Other**
<!---
Please pick one from below:
**Bug fix** (*non-breaking change which fixes an issue*)
**New feature** (*non-breaking change which adds functionality*)
**Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)
**Other** (*e.g. Documentation, Tests, Configuration*)
--->


## Description:
Remove the WARs used to keep nose alive.

nose2 supports the yield-style test discovery by default @attr has a different filtering syntax (`-A`) and just checks for presence of truthy test_foo.attribute_name. A decorator uses this mechanism for backward compatibility.

nose2 splits with_setup(setup, teardown) into two separate decorators, a backward compatible decorator is added.

nottest sets special attribute.

SkipTest from unittest is recommended to be used directly (with the same functionality).

Test scripts are adjusted with minimal changes to run through nose2. Followup cleanup can be used for renaming.

Replace unsupported -m regex by attributes

## Additional information:

### Affected modules and functionalities:
<!--- Describe here what was changed, added, removed. --->



### Key points relevant for the review:
<!--- Describe here what is the most important part that reviewers should focus on. --->

### Tests:
<!--- Describe the test coverage of the introduced change.

If you select `Existing tests apply` option, please list which test cases cover the introduced
functionality. For example:
- test_operator_gaussian_blur.py: test_gaussian*
- tensor_list_test.cc: TensorListVariableBatchSizeTest*
--->
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A


<!---
At this point you can hit "Create".
The checklist below shall be filled in the created PR.
--->

## Checklist

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

### DALI team only

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A
<!---  Introduce new or affected requirement IDs, if applicable --->

**JIRA TASK**: N/A
<!--- DALI-XXXX or NA --->

klecki committed:
Move to nose2 only

Remove the WARs used to keep nose alive.

nose2 supports the yield-style test discovery by default
@attr has a different filtering syntax (`-A`) and just checks for
presence of truthy test_foo.attribute_name. A decorator uses this
mechanism for backward compatibility.

nose2 splits with_setup(setup, teardown) into two separate decorators,
a backward compatible decorator is added.

nottest sets special attribute.

SkipTest from unittest is recommended to be used directly (with the same
functionality).

Test scripts are adjusted with minimal changes to run through nose2.
Followup cleanup can be used for renaming.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

klecki committed:
Replace unsupported -m regex by attributes

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

klecki committed:
Fix : -> .

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

klecki committed:
Greptile suggestion - use one TC instance

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

klecki committed:
Remove remaining nose

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

klecki committed:
Lint

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

mzient reviewed (commented):
The way the tests are first discovered and then run will most certainly prevent parallel external source from working (see test job 247483858) was the primary reason for not pursuing full transition to nose2 to date.
Perhaps you can move the troublesome tests to separate files to guarantee proper execution order.

Also, some of the tests used to generate vast amounts of data in the "discovery" stage, because yielding in `nose` didn't cause data accumulation. Please make sure that this is not the case any more - the tests might work when run one by one but case out-of-memory condition when multiple test files are run.

klecki committed:
Path attrib plugin to work with generators

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

klecki committed:
Fix test discovery for nose2

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

dali-automaton commented:
**CI MESSAGE**: [41872004]: BUILD STARTED

JanuszL committed:
Fixes

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

dali-automaton commented:
**CI MESSAGE**: [41872004]: BUILD FAILED

dali-automaton commented:
**CI MESSAGE**: [41914246]: BUILD FAILED

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [41949998]: BUILD FAILED

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [41958617]: BUILD FAILED

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [42935487]: BUILD PASSED

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [43115614]: BUILD FAILED

dali-automaton commented:
**CI MESSAGE**: [45742567]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [45743750]: BUILD FAILED

JanuszL committed:
Fix

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL committed:
Review fix

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL commented on dali/test/python/operator_1/test_numba_func.py:
Fixed

JanuszL commented:
!build

JanuszL committed:
Review fix

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL commented:
@greptile review

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [48772182]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [48772182]: BUILD PASSED

mzient commented on dali/test/python/test_external_source_parallel.py:
This causes instantiation of the data during test discovery. One thing to do is we should rework RandomDataIterator to create the data lazily (now it's created at construction).

JanuszL replied:
Fixed and reworked other (less significant cases)

mzient commented on dali/test/python/test_dali_tf_dataset_graph.py:
```diff
@@ -205,9 +251,15 @@ def run_tf_with_dali_external_source(dev, es_args, ed_dev, dtype, *_):
+    def setUp(self):
+        skip_inputs_for_incompatible_tf()
+
+    @params(*gen_tf_with_dali_external_source(run_tf_with_dali_external_source))
```
There's quite a bit of data materialized ahead of time. This should be reworked so that the callback is not called inside `gen_tf_with_dali_external_source`

JanuszL replied:
Fixed

mzient commented on dali/test/python/test_external_source_parallel.py:
```diff
@@ -128,45 +133,108 @@ def test_wrong_source():
+        epoch_size = 250
+        callback = utils.ExtCallback((4, 5), epoch_size, np.int32)
+        # if context is already initialized, use spawn to avoid fork wich will fail immediately
+        init_method = "fork" if not _b.IsDriverInitialized() else "spawn"
```
Wasn't it supposed to be testing `"fork"` specifically? If CUDA context is present, we could either fail or skip - but not silently fall back to the other launch method.

JanuszL replied:
Reverted this change and made calling this tests explicitly in as a separate one from the bash test.sh wrapper to make sure the context is not created beforehand

JanuszL committed:
Defer callback construction to test run time to avoid OOM at discovery

- nose_utils.py: consolidate double unittest import into single form
- test_dali_tf_es_pipelines.py: gen_tf_with_dali_external_source now
  yields a make_es_args factory instead of pre-built callback objects,
  so no data is created during test discovery
- test_dali_tf_dataset_graph.py: call make_es_args() at test run time
- test_external_source_parallel.py: _make_all_kinds_parallel_cases
  yields (epoch_size, ...) descriptors; ExtCallback/SampleCallbackBatched/
  SampleCallbackIterator are constructed inside test_all_kinds_parallel

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL commented:
@greptile review

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [48792345]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [48792345]: BUILD FAILED

JanuszL committed:
Fix _make_expected_out crash when passed a pre-built ndarray

test_numba_func_with_cond tests passed np.full(...) directly in
expected_out and were not updated when the descriptor format was
introduced in 839dfb17a, causing a ValueError on unpack.

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [48811386]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [48811386]: BUILD PASSED

mzient commented on qa/TL0_python-self-test-core/test_body.sh:
```diff
@@ -87,20 +87,20 @@ test_pytorch() {
 
         # External source tests are slow and Python-side mostly, but let's run just one of them
-        ${python_new_invoke_test} -A '!slow,!pytorch,!mxnet,!cupy,!numba' checkpointing.test_dali_checkpointing.test_external_source_checkpointing:1
+        ${python_new_invoke_test} -A '!slow,!pytorch,!cupy,!numba' checkpointing.test_dali_checkpointing.test_external_source_checkpointing.1
```
```suggestion
        ${python_new_invoke_test} -A '!slow,!pytorch,!cupy,!numba' checkpointing.test_dali_checkpointing.test_external_source_checkpointing:1
```
?

JanuszL committed:
Review fix

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

mzient commented on dali/test/python/test_dali_tf_dataset_graph.py:
```diff
@@ -205,9 +248,24 @@ def run_tf_with_dali_external_source(dev, es_args, ed_dev, dtype, *_):
+    @params(*gen_tf_with_dali_external_source())
+    def test_tf_with_dali_external_source(
+        self, get_callback, is_batched, cycle, batch_info, dense, dev, es_dev, dtype, iter_limit
+    ):
```
Why not just call this prologue in test body? We don't perform any actual setup here.
```suggestion
@params(*gen_tf_with_dali_external_source())
def test_tf_with_dali_external_source(
    self, get_callback, is_batched, cycle, batch_info, dense, dev, es_dev, dtype, iter_limit
):
    skip_inputs_for_incompatible_tf()
```

JanuszL replied:
Done

JanuszL replied:
Fixed

JanuszL committed:
Consolidate TF dataset test classes sharing setUp into single class

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

mzient commented on qa/TL0_python-self-test-core/test_body.sh:
```diff
@@ -23,19 +23,19 @@ test_py_with_framework() {
 
     if [ -z "$DALI_ENABLE_SANITIZERS" ]; then
-        ${python_new_invoke_test} -A "!slow,!pytorch,!mxnet,!cupy" test_dali_variable_batch_size
+        ${python_new_invoke_test} -A "!slow,!pytorch!cupy" test_dali_variable_batch_size
```
```suggestion
        ${python_new_invoke_test} -A "!slow,!pytorch,!cupy" test_dali_variable_batch_size
```

JanuszL replied:
Fixed

mzient commented on qa/TL0_python-self-test-core/test_body.sh:
```diff
@@ -23,19 +23,19 @@ test_py_with_framework() {
+        ${python_new_invoke_test} -A "!slow,!pytorch!cupy" test_dali_variable_batch_size
     else
-        ${python_new_invoke_test} -A "!slow,!pytorch,!mxnet,!cupy,!numba" test_dali_variable_batch_size
+        ${python_new_invoke_test} -A "!slow,!pytorch!cupy,!numba" test_dali_variable_batch_size
```
```suggestion
        ${python_new_invoke_test} -A "!slow,!pytorch,!cupy,!numba" test_dali_variable_batch_size
```

JanuszL replied:
Fixed

JanuszL committed:
Fix

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL commented on dali/test/python/test_dali_tf_dataset_eager.py:
```diff
@@ -368,30 +275,153 @@
+    with tf.device("/cpu:0"):
+        dataset = get_dali_dataset_from_pipeline(pipeline, "cpu", 0)
+
+    for sample in dataset:
```
dataset is iterable

JanuszL committed:
Add missing file

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

JanuszL committed:
Fix

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

dali-automaton commented:
**CI MESSAGE**: [49016580]: BUILD STARTED
